### PR TITLE
Initial fixes

### PR DESCRIPTION
**Pull request by juliohm**
This is a first attempt to fix some conceptual issues I encountered while reviewing the code.

We miss an implementation for `coordinates!` and an internal design choice in `interpolate` that I added in a comment. Looking at the code now I think we also need a `materializer` trait similar to `Tables.materializer` that takes both the domain and the values table as arguments and returns the original spatial data type, in this case a dimensional array.

I've also noticed that the tables interface may not make sense to dimensional array, but probably dimensional data set? The tables interface is useful when we have multiple columns of variables, and I understood that dimensional arrays contain a single layer or column. You can stack these arrays into datasets to produce a table of columns if I understand correctly.

### CODE REVIEW

**Comment by rafaqz** [comment]
Hah I just implemented that extra `coordinates!` too. 

Maybe a materializer makes sense? It's not too hard to convert back manually. What I'm missing more than that is an automatic plot recipe - it should have enough information to do that right?

And yes the table doesn't really make sense for a `DimArray`, but I wrote it so it still works as the simplest case of a single column table. Internally it's turned into a `DimDataset` with one layer first, then a `DimTable`. It's al compile time so the transformations don't really matter.

But I think your PR might be missing a lot of my code? a lot of things seem to get deleted, all those tests were working for me. Maybe do a rebase?

**Comment by rafaqz** [comment]
If you intended to remove the `RegularGrid` part, the idea is that `RegularGrid` should be faster than `DimDomain` as it gives you more information about the structure of the grid to optimize on.

Does that not make any difference?

File: dimdomain.jl [hunk 1]
```diff
@@ -1,171 +1,88 @@
...
+Either interpolate to fill missings in the current array,
 or interpolate to a new domain specified by `dims`.
-
-If the passed in object is a `AbstractDimDataset`, all layers will be interpolated.
 """
 interpolate(da::AbstractDimArray, solver::AbstractSolver; dims=dims(da)) = begin
-    key = name(da)
-    data = SpatialData(domain(da), DimTable(da)) 
-    problem = EstimationProblem(data, domain(dims), key, mapper=CopyMapper())
-    sol = solve(problem, solver)
```
Comment (rafaqz) [inline dimdomain.jl:90]: Why is all the interpolation code deleted?

**Comment by juliohm** [comment]
Yes, I commented out some of the tests because they were conceptually incorrect. For example, you cannot simulate on a dimensional array (because it implements the data trait). You can simulate on domain objects though `domain(da)`.

Regarding the optimization with `RegularGrid`, I thought we could think about these details later after we have a MWE. The idea would be to use external types first, to make sure everything works as expected, and then think about integrations and where things should go. If we start to use `RegularGrid` we will add a dependency on GeoStatsBase.jl right away.

**Comment by juliohm** [comment]
Regarding the plot recipe, we have a fallback implementation for spatial data on unknown domain types, which is basically the same plot you would get for a point set. However, if `DimDomain` implements the correct plot recipe following our conventions, it should just work. You can see examples in the `plotrecipes` folder in GeoStatsbase.jl where we specialize the plots for RegularGrid

**Comment by rafaqz** [comment]
This PR basically trashes a working implementation with tests. I'm confused. Why have you deleted so much working code.

The conceptual problem should be a one line fix??

**Comment by juliohm** [comment]
I think the code was too crowded with details that were not relevant to get a dimensional array working with the spatial data trait. All you need is to provide two functions as in this PR. It seems that the `DimTable` is already in DimensionalData.jl, so we can just forward `GeoStatsBase.values(da) = DimTable(da)`. This PR just adds the missing part `DimDomain`, and forwards it similarly: `GeoStatsBase.domain(da) = DimDomain(da)`. That should the MWE, right?

**Comment by rafaqz** [comment]
But this version now does not work.

You can't clean up a working version with a broken version.

**Comment by juliohm** [comment]
Well, the missing element is `coordinates!` after that things should work?

File: dimdomain.jl [hunk 2]
```diff
@@ -1,171 +1,88 @@
...
-
-# Domain interface for DimDomain
-
-GeoStatsBase.nelms(dom::DimDomain) = length(first(dimcolumns(dom)))
-GeoStatsBase.ncoords(dom::DimDomain) = length(dimcolumns(dom))
-GeoStatsBase.coordtype(dom::DimDomain) = eltype(first(dimcolumns(dom)))
-
-# Why do we need both of these methods?
-GeoStatsBase.coordinates!(buf::AbstractVector, dom::DimDomain, i::Int) = 
-    buf[i] = SA[map(c -> c[i], dimcolumns(dom))...]
```
Comment (rafaqz) [inline dimdomain.jl:30]: these are the coordinates methods why are they deleted?

File: dimdomain.jl [hunk 3]
```diff
@@ -1,171 +1,88 @@
...
+GeoStatsBase.ncoords(dom::DimDomain) = length(dims(dom.dimcolumns))
 
+# this should return the type of the lat/lon pairs, probably Float64
+# I will just hard code it for now
+GeoStatsBase.coordtype(dom::DimDomain) = Float64
 
-# Data interface
+# this should set the buf to contain the coordinates of the i-th cell
+# in the dimensional array
+GeoStatsBase.coordinates!(buf, dom::DimDomain, i) = @error "not implemented"
```
Comment (rafaqz) [inline dimdomain.jl:34]: and why replaced with a "not implemented" method? there is a working version above this??

**Comment by juliohm** [comment]
I couldn't reason clearly with so much code implementing traits that were not supposed to be implemented. If you can provide the missing `coordinates!` and run, then we can start investigating what may be failing in the tests. Most tests were conceptually flawed, that is the point I am trying to raise here.

File: dimdomain.jl (hunk 1 above)
Comment (rafaqz) [reply dimdomain.jl:90]: don't delete a working example without a replacement or fix? who is meant to write that?

**Comment by rafaqz** [comment]
This PR deletes working code and adds nothing of value

**Comment by rafaqz** [comment]
Please write actual working code to contribute something here

**Comment by juliohm** [comment]
Ok, I tried but it is hard to convince you that you are getting the concepts wrong.

**Comment by rafaqz** [comment]
No you didn't... you just deleted a bunch of things. I'm totally open to hearing how I'm getting things wrong.

The best thing to convince me is a working demo :)

**Comment by rafaqz** [comment]
But why did you delete the coordinates methods??? I'm seriously so confused.

**Comment by rafaqz** [comment]
Where is your version of interpolate? we need an implementation that actually does something useful.

**Comment by rafaqz** [comment]
This is really bizzarre. I really can't understand why you would try to communicate by deleting everything, instead of just talking through the problems and things I don't understand in an issue, or proposing a working alternative.

**Comment by juliohm** [comment]
@rafaqz have you heard of the proverb "less is more"? You had a lot of code doing the wrong things. I erased that code and added the minimum amount of code doing the right things (because I am the author of the traits and I know what to do?). You then started to complain about previous tests erased when these same tests were completely off conceptually. What else can I do? I am not feeling that this is a healthy collaboration, and I really think you should reassess your past collaborations to see if you had similar issues with your ego.

Anyways, let's both take some time off because it is Sunday. I am going offline for a bit.

**Comment by rafaqz** [comment]
But you deleted the coordinates methods and replaced them with a broken method? Why?

And the interpolate method that was an actual useful demo. It was a basic copy of the working implementation in GeoArrays.jl... maybe I messed something up but it's a small fix. Interpolation is the main thing I actually need out of this whole effort.

And the `ReguarGrid` actually makes sense. I understand the performance trade-offs here, why would you delete that? `DimDomain` indexing is fundamentally slower.

**Comment by rafaqz** [comment]
I am also not feeling its a healthy collaboration. I write all the code, you just delete some occasionally.

I'm going to make this a separate glue package, feel free to use it how you like.

**Comment by rafaqz** [comment]
The interpolator was copied off GeoArrays.jl. It's their working implementation. Would you go and delete that in their repo to?

What it does by default is it fills in `missing` values. That's useful in itself, and all it does in GeoArrays.jl

If you pass in a larger dims tuple to dims, it interpolates to a larger grid. It seems to work?


Did you actually run it?

File: dimdomain.jl [hunk 4]
```diff
@@ -1,171 +1,88 @@
...
-    problem = EstimationProblem(data, domain(dims), key, mapper=CopyMapper())
-    sol = solve(problem, solver)
-    newA = reshape(sol[key][:mean], map(length, dims))
-    newdims = formatdims(newA, dims)
-    rebuild(da, newA, newdims)
+    # it seems that an abstract array stores a single layer?
+    # in this case, get the layer name, and interpolate on
+    # a possibly finer grid. how do we construct such grid?
+    # ideally we should be able to easily construct a DimDomain
+    # from the dims keyword argument. because I am not familiar
```
Comment (rafaqz) [inline dimdomain.jl:52]: Constructing possibly finer grid is what the code you deleted already did. The dims arg changes the grid size.

Deleted peoples working code is no way to collaborate with them. If you don't understand how DD works why did you delete this implementation?

**Comment by rafaqz** [comment]
FYI this is the result of the `interpolation` method you said didn't work:

Initial random array:
original

Interpolated finer resolution array, by passing in a `dims` tuple with smaller step size in the axes, as in the tests:
interpolated

It seems to work pretty well to me. You can plot it for yourself in the tests.
